akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -0,0 +1,297 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery
+
+import java.nio.charset.Charset
+import java.nio.{ ByteBuffer, ByteOrder }
+
+import org.agrona.concurrent.{ ManyToManyConcurrentArrayQueue, UnsafeBuffer }
+
+/**
+ * INTERNAL API
+ */
+private[remote] class OutOfBuffersException extends RuntimeException("Out of usable ByteBuffers")
+
+/**
+ * INTERNAL API
+ */
+private[remote] class EnvelopeBufferPool(maximumPayload: Int, maximumBuffers: Int) {
+  private val availableBuffers = new ManyToManyConcurrentArrayQueue[EnvelopeBuffer](maximumBuffers)
+
+  def acquire(): EnvelopeBuffer = {
+    val buf = availableBuffers.poll()
+    if (buf ne null) {
+      buf
+    } else {
+      val newBuf = new EnvelopeBuffer(ByteBuffer.allocateDirect(maximumPayload))
+      newBuf.byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+      newBuf
+    }
+  }
+
+  def release(buffer: EnvelopeBuffer) = {
+    // If pool is full, this will just drop, forcing GC to collect the extra buffer
```

Comment:
isn't that what https://github.com/akka/akka/issues/17194 is about?
using `sun.misc.Cleaner`